crates/bevy_ecs/src/schedule/schedule.rs

```diff
@@ -863,57 +859,80 @@ impl ScheduleGraph {
         components: &Components,
     ) -> Result<SystemSchedule, ScheduleBuildError> {
         // check hierarchy for cycles
-        self.hierarchy.topsort = self
-            .topsort_graph(&self.hierarchy.graph, ReportCycles::Hierarchy)
-            .map_err(|_| ScheduleBuildError::HierarchyCycle)?;
+        self.hierarchy.topsort =
+            self.topsort_graph(&self.hierarchy.graph, ReportCycles::Hierarchy)?;
 
         let hier_results = check_graph(&self.hierarchy.graph, &self.hierarchy.topsort);
-        if self.settings.hierarchy_detection != LogLevel::Ignore
-            && self.contains_hierarchy_conflicts(&hier_results.transitive_edges)
-        {
-            self.report_hierarchy_conflicts(&hier_results.transitive_edges);
-            if matches!(self.settings.hierarchy_detection, LogLevel::Error) {
-                return Err(ScheduleBuildError::HierarchyRedundancy);
-            }
-        }
+        self.optionally_check_hierarchy_conflicts(&hier_results.transitive_edges)?;
 
         // remove redundant edges
         self.hierarchy.graph = hier_results.transitive_reduction;
 
         // check dependencies for cycles
-        self.dependency.topsort = self
-            .topsort_graph(&self.dependency.graph, ReportCycles::Dependency)
-            .map_err(|_| ScheduleBuildError::DependencyCycle)?;
+        self.dependency.topsort =
+            self.topsort_graph(&self.dependency.graph, ReportCycles::Dependency)?;
 
         // check for systems or system sets depending on sets they belong to
         let dep_results = check_graph(&self.dependency.graph, &self.dependency.topsort);
-        for &(a, b) in dep_results.connected.iter() {
-            if hier_results.connected.contains(&(a, b)) || hier_results.connected.contains(&(b, a))
-            {
-                let name_a = self.get_node_name(&a);
-                let name_b = self.get_node_name(&b);
-                return Err(ScheduleBuildError::CrossDependency(name_a, name_b));
-            }
-        }
+        self.check_for_cross_dependencies(&dep_results, &hier_results.connected)?;
 
         // map all system sets to their systems
         // go in reverse topological order (bottom-up) for efficiency
+        let (set_systems, set_system_bitsets) =
+            self.map_sets_to_systems(&self.hierarchy.topsort, &self.hierarchy.graph);
+        self.check_order_but_intersect(&dep_results.connected, &set_system_bitsets)?;
+
+        // check that there are no edges to system-type sets that have multiple instances
+        self.check_system_type_set_ambiguity(&set_systems)?;
+
+        // flatten: combine `in_set` with `before` and `after` information
+        // have to do it like this to preserve transitivity
+        let dependency_flattened = self.get_dependency_flattened(&set_systems);
```

Comment:
I believe this comment should go inside of the `get_dependency_flattened` function.

Reply:
@hymm I'm going to wait to merge until this is fixed up, but I'll merge ASAP once it is.